File: app/src/main/java/net/veldor/flibustaloader/utils/UriConverter.java
package net.veldor.flibustaloader.utils;

import android.content.Context;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.RequiresApi;

import java.io.File;

public class UriConverter {
    /**
     * Ing.N.Nyerges 2019 V2.0
     *
     * Storage Access Framework(SAF) Uri's creator from File (java.IO),
     * for removable external storages
     *
     * @param context Application Context
     * @param file File path + file name
     * @return Uri[]:
     *   uri[0] = SAF TREE Uri
     *   uri[1] = SAF DOCUMENT Uri
     */
    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    public static Uri[] getSafUris (Context context, File file) {

        Uri[] uri = new Uri[2];
        String scheme = "content";
        String authority = "com.android.externalstorage.documents";

        // Separate each element of the File path
        // File format: "/storage/XXXX-XXXX/sub-folder1/sub-folder2..../filename"
        // (XXXX-XXXX is external removable number
        String[] ele = file.getPath().split(File.separator);
        //  ele[0] = not used (empty)
        //  ele[1] = not used (storage name)
        //  ele[2] = storage number
        //  ele[3 to (n-1)] = folders
        //  ele[n] = file name

        // Construct folders strings using SAF format
        StringBuilder folders = new StringBuilder();
        if (ele.length > 4) {
            folders.append(ele[3]);
            for (int i = 4; i < ele.length - 1; ++i) folders.append("%2F").append(ele[i]);
        }

        String common = ele[2] + "%3A" + folders.toString();

        // Construct TREE Uri
        Uri.Builder builder = new Uri.Builder();
        builder.scheme(scheme);
        builder.authority(authority);
        builder.encodedPath("/tree/" + common);
        uri[0] = builder.build();

        // Construct DOCUMENT Uri
        builder = new Uri.Builder();
        builder.scheme(scheme);
        builder.authority(authority);
        if (ele.length > 4) common = common + "%2F";
        builder.encodedPath("/document/" + common + file.getName());
        uri[1] = builder.build();

        return uri;
    }
}

File: app/src/main/java/net/veldor/flibustaloader/utils/UriConverter.kt
package net.veldor.flibustaloader.utils

import android.content.Context
import android.os.Build
import android.provider.DocumentsContract
import android.os.Environment
import androidx.annotation.RequiresApi
import android.net.Uri

internal object UriConverter {
    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    fun getPath(context: Context?, uri: Uri): String? {
        val isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT

        // DocumentProvider
        if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
            println("getPath() uri: $uri")
            println("getPath() uri authority: " + uri.authority)
            println("getPath() uri path: " + uri.path)

            // ExternalStorageProvider
            if ("com.android.externalstorage.documents" == uri.authority) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":").toTypedArray()
                val type = split[0]
                println("getPath() docId: " + docId + ", split: " + split.size + ", type: " + type)

                // This is for checking Main Memory
                return if ("primary".equals(type, ignoreCase = true)) {
                    if (split.size > 1) {
                        Environment.getExternalStorageDirectory().toString() + "/" + split[1] + "/"
                    } else {
                        Environment.getExternalStorageDirectory().toString() + "/"
                    }
                    // This is for checking SD Card
                } else {
                    "storage" + "/" + docId.replace(":", "/")
                }
            }
        }
        return null
    }
}